# 연습문제 2-2

세 부분계 $A_{1},A_{2},A_{3}$ 로 이루어진 결합계 $A_{0}$가 있다. 각 부분계의 자유도 수는 각각 $f_{1}=2, f_{2}=4, f_{3}=8$이고, 에너지는 각각 $E_{1}, E_{2}, E_{3}$이다. 결합계의 총 에너지가 $E_{0}$일 떄, 결합계의 모든 에너지분포에 대한 상태수를 구하고, 가장 가능한 에너지분포의 상태수가 차지하는 비율을 구하여라

python의 data analysis package인 pandas를 이용해서 표를 그려보자!

In [1]:
from pandas import Series, DataFrame
import pandas as pd
import numpy as np

In [2]:
f1,f2,f3=2,4,8
data={'Degree of freedom': [f1,f2,f3]}
frame=DataFrame(data,index=['A1','A2','A3'])# 간단한 데이터 프레임
frame

Degree of freedom
A1                  2
A2                  4
A3                  8

$\Omega(E)=E^{f/2}$ 라고 가정하고 문제를 풀자. 이제 for loop를 이용해서 $(E_1,E_2,E_3,\Omega_1,\Omega_2,\Omega_3,\Omega_0)$ 세트들을 구하자

In [3]:
E0=4
Emat=np.array([[0,0,0,0,0,0,0]]) #더미 세트: 더미 세트를 없앨 수 있는 방법은?

for i in range(E0+1):
    E1=i
    for j in range(E0-E1+1):
        E2=j
        E3=E0-(E1+E2)
        Omega1=E1**(f1/2)
        Omega2=E2**(f2/2)
        Omega3=E3**(f3/2)
        Omega0=Omega1*Omega2*Omega3
        Emat=np.append(Emat,[[E1,E2,E3,Omega1,Omega2,Omega3,Omega0]],axis=0)
        
Emat

array([[  0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [  0.,   0.,   4.,   0.,   0., 256.,   0.],
       [  0.,   1.,   3.,   0.,   1.,  81.,   0.],
       [  0.,   2.,   2.,   0.,   4.,  16.,   0.],
       [  0.,   3.,   1.,   0.,   9.,   1.,   0.],
       [  0.,   4.,   0.,   0.,  16.,   0.,   0.],
       [  1.,   0.,   3.,   1.,   0.,  81.,   0.],
       [  1.,   1.,   2.,   1.,   1.,  16.,  16.],
       [  1.,   2.,   1.,   1.,   4.,   1.,   4.],
       [  1.,   3.,   0.,   1.,   9.,   0.,   0.],
       [  2.,   0.,   2.,   2.,   0.,  16.,   0.],
       [  2.,   1.,   1.,   2.,   1.,   1.,   2.],
       [  2.,   2.,   0.,   2.,   4.,   0.,   0.],
       [  3.,   0.,   1.,   3.,   0.,   1.,   0.],
       [  3.,   1.,   0.,   3.,   1.,   0.,   0.],
       [  4.,   0.,   0.,   4.,   0.,   0.,   0.]])

pandas의 데이터프레임으로 구현해보자

In [4]:
frame1=DataFrame(Emat,columns=['E1','E2','E3','Omega1','Omega2','Omega3','Omega0'])
frame1

E1   E2   E3  Omega1  Omega2  Omega3  Omega0
0   0.0  0.0  0.0     0.0     0.0     0.0     0.0
1   0.0  0.0  4.0     0.0     0.0   256.0     0.0
2   0.0  1.0  3.0     0.0     1.0    81.0     0.0
3   0.0  2.0  2.0     0.0     4.0    16.0     0.0
4   0.0  3.0  1.0     0.0     9.0     1.0     0.0
5   0.0  4.0  0.0     0.0    16.0     0.0     0.0
6   1.0  0.0  3.0     1.0     0.0    81.0     0.0
7   1.0  1.0  2.0     1.0     1.0    16.0    16.0
8   1.0  2.0  1.0     1.0     4.0     1.0     4.0
9   1.0  3.0  0.0     1.0     9.0     0.0     0.0
10  2.0  0.0  2.0     2.0     0.0    16.0     0.0
11  2.0  1.0  1.0     2.0     1.0     1.0     2.0
12  2.0  2.0  0.0     2.0     4.0     0.0     0.0
13  3.0  0.0  1.0     3.0     0.0     1.0     0.0
14  3.0  1.0  0.0     3.0     1.0     0.0     0.0
15  4.0  0.0  0.0     4.0     0.0     0.0     0.0

Omega0 열들을 따로 빼서 비율을 계산한다.

In [5]:
frame1.Omega0  # 이렇게 따로 빼면 독립적인 column처럼 취급할 수 있다.

0      0.0
1      0.0
2      0.0
3      0.0
4      0.0
5      0.0
6      0.0
7     16.0
8      4.0
9      0.0
10     0.0
11     2.0
12     0.0
13     0.0
14     0.0
15     0.0
Name: Omega0, dtype: float64

In [6]:
irange=int(len(frame1.Omega0)) # =16
irange

16

In [7]:
tot=np.sum(frame1.Omega0)
percentage=[]

for i in range(irange):
    percentage.append(100*frame1.Omega0[i]/tot)
percentage

[0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 72.72727272727273,
 18.181818181818183,
 0.0,
 0.0,
 9.090909090909092,
 0.0,
 0.0,
 0.0,
 0.0]

In [8]:
frame1=DataFrame(Emat,percentage,columns=['E1','E2','E3','Omega1','Omega2','Omega3','Omega0'])
frame1

E1   E2   E3  Omega1  Omega2  Omega3  Omega0
0.000000   0.0  0.0  0.0     0.0     0.0     0.0     0.0
0.000000   0.0  0.0  4.0     0.0     0.0   256.0     0.0
0.000000   0.0  1.0  3.0     0.0     1.0    81.0     0.0
0.000000   0.0  2.0  2.0     0.0     4.0    16.0     0.0
0.000000   0.0  3.0  1.0     0.0     9.0     1.0     0.0
0.000000   0.0  4.0  0.0     0.0    16.0     0.0     0.0
0.000000   1.0  0.0  3.0     1.0     0.0    81.0     0.0
72.727273  1.0  1.0  2.0     1.0     1.0    16.0    16.0
18.181818  1.0  2.0  1.0     1.0     4.0     1.0     4.0
0.000000   1.0  3.0  0.0     1.0     9.0     0.0     0.0
0.000000   2.0  0.0  2.0     2.0     0.0    16.0     0.0
9.090909   2.0  1.0  1.0     2.0     1.0     1.0     2.0
0.000000   2.0  2.0  0.0     2.0     4.0     0.0     0.0
0.000000   3.0  0.0  1.0     3.0     0.0     1.0     0.0
0.000000   3.0  1.0  0.0     3.0     1.0     0.0     0.0
0.000000   4.0  0.0  0.0     4.0     0.0     0.0     0.0

가장 가능한 에너지 분포의 상태수는 (1,1,2) 이고 비율은 72.727273% 이다. 따라서 이 상태로 결합될 확률은 72.727273%!

# 연습문제 2-3

한편, $f_{1}=2\times10^{24}, f_{2}=4\times10^{24}, f_{3}=8\times10^{24}$ 일 때, 위의 과정을 되풀이해보면,

In [15]:
f1,f2,f3=2e24,4e24,8e24

E0=4
Emat2=np.array([[0,0,0,0,0,0,0]]) #더미 세트: 더미 세트를 없앨 수 있는 방법은?

for i in range(E0+1):
    E1=i
    for j in range(E0-E1+1):
        E2=j
        E3=E0-(E1+E2)
        Omega1=E1**(f1/2)
        Omega2=E2**(f2/2)
        Omega3=E3**(f3/2)
        Omega0=Omega1*Omega2*Omega3
        Emat2=np.append(Emat2,[[E1,E2,E3,Omega1,Omega2,Omega3,Omega0]],axis=0) #Emat에 항목들 추가하기
        
frame2=DataFrame(Emat2,columns=['E1','E2','E3','Omega1','Omega2','Omega3','Omega0']) #데이터프레임 정의하기

tot=np.sum(frame2.Omega0) #Omega0 element 들의 총 합
irange=len(frame2.Omega0) #Omega0 element의 갯수
percentage=[]  #percentage를 담을 empty list

for i in range(irange):
    percentage.append(100*frame2.Omega0[i]/tot)

frame2=DataFrame(Emat2,percentage,columns=['E1','E2','E3','Omega1','Omega2','Omega3','Omega0'])

OverflowError: (34, 'Result too large')

숫자가 너무 커서 계산을 안한다.

In [18]:
f1,f2,f3=200,400,800

E0=4
Emat2=np.array([[0,0,0,0,0,0,0]]) #더미 세트: 더미 세트를 없앨 수 있는 방법은?

for i in range(E0+1):
    E1=i
    for j in range(E0-E1+1):
        E2=j
        E3=E0-(E1+E2)
        Omega1=E1**(f1/2)
        Omega2=E2**(f2/2)
        Omega3=E3**(f3/2)
        Omega0=Omega1*Omega2*Omega3
        Emat2=np.append(Emat2,[[E1,E2,E3,Omega1,Omega2,Omega3,Omega0]],axis=0) #Emat에 항목들 추가하기
        
frame2=DataFrame(Emat2,columns=['E1','E2','E3','Omega1','Omega2','Omega3','Omega0']) #데이터프레임 정의하기

tot=np.sum(frame2.Omega0) #Omega0 element 들의 총 합
irange=len(frame2.Omega0) #Omega0 element의 갯수
percentage=[]  #percentage를 담을 empty list

for i in range(irange):
    percentage.append(100*frame2.Omega0[i]/tot)

frame2=DataFrame(Emat2,percentage,columns=['E1','E2','E3','Omega1','Omega2','Omega3','Omega0'])
frame2

E1   E2   E3        Omega1         Omega2         Omega3  \
0.000000e+00  0.0  0.0  0.0  0.000000e+00   0.000000e+00   0.000000e+00   
0.000000e+00  0.0  0.0  4.0  0.000000e+00   0.000000e+00  6.668014e+240   
0.000000e+00  0.0  1.0  3.0  0.000000e+00   1.000000e+00  7.055079e+190   
0.000000e+00  0.0  2.0  2.0  0.000000e+00   1.606938e+60  2.582250e+120   
0.000000e+00  0.0  3.0  1.0  0.000000e+00   2.656140e+95   1.000000e+00   
0.000000e+00  0.0  4.0  0.0  0.000000e+00  2.582250e+120   0.000000e+00   
0.000000e+00  1.0  0.0  3.0  1.000000e+00   0.000000e+00  7.055079e+190   
1.000000e+02  1.0  1.0  2.0  1.000000e+00   1.000000e+00  2.582250e+120   
6.223015e-59  1.0  2.0  1.0  1.000000e+00   1.606938e+60   1.000000e+00   
0.000000e+00  1.0  3.0  0.0  1.000000e+00   2.656140e+95   0.000000e+00   
0.000000e+00  2.0  0.0  2.0  1.267651e+30   0.000000e+00  2.582250e+120   
4.909093e-89  2.0  1.0  1.0  1.267651e+30   1.000000e+00   1.000000e+00   
0.000000e+00  2.0  2.0  0.0  1.267651e+30   1.606938e+60   0.000000e+00   
0.000000e+00  3.0  0.0  1.0  5.153775e+47   0.000000e+00   1.000000e+00   
0.000000e+00  3.0  1.0  0.0  5.153775e+47   1.000000e+00   0.000000e+00   
0.000000e+00  4.0  0.0  0.0  1.606938e+60   0.000000e+00   0.000000e+00   

                     Omega0  
0.000000e+00   0.000000e+00  
0.000000e+00   0.000000e+00  
0.000000e+00   0.000000e+00  
0.000000e+00   0.000000e+00  
0.000000e+00   0.000000e+00  
0.000000e+00   0.000000e+00  
0.000000e+00   0.000000e+00  
1.000000e+02  2.582250e+120  
6.223015e-59   1.606938e+60  
0.000000e+00   0.000000e+00  
0.000000e+00   0.000000e+00  
4.909093e-89   1.267651e+30  
0.000000e+00   0.000000e+00  
0.000000e+00   0.000000e+00  
0.000000e+00   0.000000e+00  
0.000000e+00   0.000000e+00

자유도가 200,400,800만 되어도 (1,1,2)으로 결합할 확률이 100퍼센트로 나온다. 따라서 자유도가 아보가드로 단위라면 사실상 100퍼센트라고 생각해도 무방할 것이다.